File: LeetCode/Tree/TraverseTree/BFS/Leetcode_2641_cousins_in_binary_tree_ii/Leetcode_2641_cousins_in_binary_tree_ii.cpp
/**
 * @brief: Leetcode_2641_二叉树的堂兄弟节点_II
 * @link: https://leetcode.cn/problems/cousins-in-binary-tree-ii/description/
 * @author: liangj.zhang
 * @date: 11/1/2026
 * 
 * @Difficulty: Medium
 * 
 * @Label: BFS
 * 
 * @Retrospect(worthy 1 - 5): 3
 * 
 * @thoughts:
 *  + 【思路 1】：BFS + vector
 *      存入一个节点时，就把该节点的孩子节点和传下去；
 *      同时在这个过程，也能算出下一层的总和； => 不用在下一层节点遍历时，额外遍历一遍；
 *      + 分析：
 *          + 时间复杂度：O(n)
 *          + 空间复杂度：O(n)
 *      + rank:
 *          + 时间效率：7 ms, 击败 96.95%
 *          + 空间效率：294.92 MB, 击败 99.49%
 */

#include "../../../../../Include/Leetcode/Tree/Tree.h"
using Leetcode::Tree::BinaryTree::TreeNode;
using namespace Leetcode::Tree::BinaryTree;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <tuple>
using std::tuple;

class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        
        vector<tuple<TreeNode*, int>> arr{{root, root->val}}, tmp;  // 这里 root->val 表示，兄弟节点之和；
        int last_sum = root->val;   // 假定上一层之和；

        while (arr.size()) {

            int new_sum = 0;
            tmp.clear();    // 原来错误的写法，忘记这里的 clear() 
            for (auto [cur, sub_sum] : arr) {

                int _sum = (cur->left ? cur->left->val : 0) + 
                                    (cur->right ? cur->right->val : 0);
                if (cur->left) tmp.push_back({cur->left, _sum});
                if (cur->right) tmp.push_back({cur->right, _sum});
                cur->val = last_sum - sub_sum;
                new_sum += _sum;
            }
            arr.swap(tmp);
            last_sum = new_sum;
        }

        return root;
    }
};

int main() {

    TreeNode* root = buildTreeFromStringArray("[5,4,9,1,10,null,7]");
    TreeNode* res = Solution().replaceValueInTree(root);
    levelTraverse(res);
}

File: LeetCode_Python/Tree/TraverseTree/BFS/Leetcode_2641_cousins_in_binary_tree_ii/Leetcode_2641_cousins_in_binary_tree_ii.py
"""
@brief: Leetcode_2641_二叉树的堂兄弟节点_II
@link: https://leetcode.cn/problems/cousins-in-binary-tree-ii/description/
@author: liangj.zhang
@date: 11/1/2026

@Difficulty: Medium

@Label: BFS

@Retrospect(worthy 1 - 5): 3

@thoughts:
 + 【思路 1】：BFS + vector
     存入一个节点时，就把该节点的孩子节点和传下去；
     同时在这个过程，也能算出下一层的总和； => 不用在下一层节点遍历时，额外遍历一遍；
     + 分析：
         + 时间复杂度：O(n)
         + 空间复杂度：O(n)
     + rank:
         + 时间效率：95 ms, 击败 65.37%
         + 空间效率：63.39 MB, 击败 12.99%
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        arr, tmp = [(root, root.val)], []
        last_sum = root.val

        while arr.__len__():
            new_sum = 0
            tmp = []    # 这里不能是 tmp.clear(), arr = tmp，arr 和 tmp 是同一个引用，清空也会将 arr 的指向内容清空
            for cur, sub_sum in arr:
                _sum = (cur.left.val if cur.left else 0) + \
                        (cur.right.val if cur.right else 0)
                if cur.left: tmp.append((cur.left, _sum))
                if cur.right: tmp.append((cur.right, _sum))
                cur.val = last_sum - sub_sum
                new_sum += _sum
            arr = tmp
            last_sum = new_sum

        return root